### PR TITLE
Fix PowerPC specific issues

### PR DESCRIPTION
**Pull request by LinHu2016**
-Replace `do_{_if_(condition)_{_..._continute;_}_}while_{false};` with
  `for(;;)_{_if_(condition)_{...}_else{break;_}_}` for an infinite loop
  with a break logic to handle waiting concurrent gc finish during
  continuation mounting. compiler might generate wrong logic for
  do_while infinite loop.

 -fix small hole during mounting synchronization with concurrent
  scanning(for the case concurrent scanning start just before resetting
  pendingmount flag-before reading the value of state for atomic update)

- new assertion check for catching issue earlier.

fix: https://github.com/eclipse-openj9/openj9/issues/16728#issuecomment-1843799810 [ISSUE3]

Signed-off-by: hulin <linhu@ca.ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
jenkins compile amac jdk21

File: runtime/vm/ContinuationHelpers.cpp [hunk 1]
```diff
@@ -148,15 +148,16 @@ createContinuation(J9VMThread *currentThread, j9object_t continuationObject)
...
 
 j9object_t
-synchronizeWithConcurrentGCScan(J9VMThread *currentThread, j9object_t continuationObject, ContinuationState volatile *continuationStatePtr)
+synchronizeWithConcurrentGCScan(J9VMThread *currentThread, j9object_t continuationObject, volatile ContinuationState *continuationStatePtr)
 {
-	ContinuationState oldContinuationState = 0;
-	ContinuationState returnContinuationState = 0;
+	volatile ContinuationState oldContinuationState = 0;
+	volatile ContinuationState returnContinuationState = 0;
+	volatile ContinuationState newContinuationState = 0;
```
Comment (amicic) [inline runtime/vm/ContinuationHelpers.cpp:155]: I'd remove volatile here, and in other methods where used for local variables (not for a pointer to a volatile location)

File: runtime/gc_base/GCExtensions.cpp [hunk 2]
```diff
@@ -319,7 +319,7 @@ MM_GCExtensions::needScanStacksForContinuationObject(J9VMThread *vmThread, j9obj
 {
 	bool needScan = false;
 #if JAVA_SPEC_VERSION >= 19
-	ContinuationState volatile *continuationStatePtr = VM_ContinuationHelpers::getContinuationStateAddress(vmThread, objectPtr);
+	volatile ContinuationState *continuationStatePtr = VM_ContinuationHelpers::getContinuationStateAddress(vmThread, objectPtr);
```
Comment (amicic) [inline runtime/gc_base/GCExtensions.cpp:322]: while I prefer the new style, the old style should be still valid with identical effects - so I'd minimize the change and revert to the old style

**Comment by pshipton** [comment]
jenkins compile aix jdk21

File: runtime/vm/ContinuationHelpers.cpp [hunk 3]
```diff
@@ -167,10 +168,17 @@ synchronizeWithConcurrentGCScan(J9VMThread *currentThread, j9object_t continuati
 	 * but only up to small finite number of times.
 	 */
 	do {
+		if (oldContinuationState == returnContinuationState) {
+			oldContinuationState = newContinuationState;
+
+		} else {
+			oldContinuationState = returnContinuationState;
+		}
```
Comment (amicic) [inline runtime/vm/ContinuationHelpers.cpp:176]: this is hard to follow
either we need more comments or better we just start this loop from scratch: read oldContinuationState from continuationStatePtr and check isConcurrentlyScanned against oldContinuationState...

**Comment by amicic** [comment]
The change looks good now (assuming it still fixed the problem), although I have hard time believing that the part that changed _while_ to _for_ loop had any effect. We'll have to investigate understand that part more...

**Comment by amicic** [comment]
jenkins test sanity aix,xLinux,win jdk21

**Comment by amicic** [comment]
jenkins test sanity win jdk21

**Comment by pshipton** [comment]
Windows machines don't seem to be working too well atm.

**Comment by pshipton** [comment]
jenkins test sanity xmac jdk21

File: runtime/vm/ContinuationHelpers.cpp [hunk 4]
```diff
@@ -183,20 +185,21 @@ synchronizeWithConcurrentGCScan(J9VMThread *currentThread, j9object_t continuati
...
 		/* Remove pending state */
-		oldContinuationState = *continuationStatePtr;
 		Assert_VM_true(VM_ContinuationHelpers::isMountedWithCarrierThread(oldContinuationState, currentThread));
 		Assert_VM_true(VM_ContinuationHelpers::isPendingToBeMounted(oldContinuationState));
 		ContinuationState newContinuationState = oldContinuationState;
 		VM_ContinuationHelpers::resetPendingState(&newContinuationState);
 		returnContinuationState = VM_AtomicSupport::lockCompareExchange(continuationStatePtr, oldContinuationState, newContinuationState);
 	} while (oldContinuationState != returnContinuationState);
 	Assert_VM_false(VM_ContinuationHelpers::isPendingToBeMounted(*continuationStatePtr));
+	Assert_VM_false(VM_ContinuationHelpers::isConcurrentlyScanned(*continuationStatePtr));
```
Comment (amicic) [inline runtime/vm/ContinuationHelpers.cpp:202]: Actually, I think this assert is ambitious since it's based on newly re-read state. Another scan can start just after this one finished. It could be a rescan within the same cycle (concurrent card cleaning) or in another cycle of same type or overlapping cycle of the complement type.

**Comment by amicic** [comment]
While  windows and mac sanity testing failed,  I don't see any real functional problems.

**Comment by pshipton** [comment]
Pls cherry pick for the 0.42 and 0.43 branches.
